lib/update/cluster/steps.go

```diff
@@ -43,13 +43,14 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-func (r *phaseBuilder) initSteps(ctx context.Context) error {
+func (r *phaseBuilder) initSteps(ctx context.Context) (err error) {
+	var installedOrUpgradedEtcdVersion *semver.Version
+
 	if !r.isDirectUpgrade() {
-		steps, err := r.buildIntermediateSteps(ctx)
+		r.steps, installedOrUpgradedEtcdVersion, err = r.buildIntermediateSteps(ctx)
```

Comment:
I think this might be confusing overall and I'd consider moving the logic of extracting the last etcd version to a separate function:
```go
func (r phaseBuilder) getInstalledOrUpgradedEtcdVersion() (result *semver.Version, err error) {
	if len(r.steps) != 0 {
		if lastStep := r.steps[len(r.steps)-1].etcd; lastStep != nil {
			return semver.New(lastStep.update), nil
		}
	}
	result, err = getEtcdVersionFromManifest(r.installedApp.Manifest, r.packages)
	if err != nil && !trace.IsNotFound(err) {
		return nil, trace.Wrap(err)
	}
	if result == nil {
		result = &r.currentEtcdVersion
	}
	return result, nil
}
```
and then:
```go
if !r.isDirectUpgrade() {
	r.steps, err = r.buildIntermediateSteps(ctx)
	if err != nil {
		return trace.Wrap(err)
	}
}
installedOrUpgradedEtcdVersion, err := r.getInstalledOrUpgradedEtcdVersion()
// ...
```
Just my two cents.

Reply:
Yea I agree it's confusing and did consider going an approach with a separate function. My biggest worry is it creates an opportunity to break by getting out of sync with future changes to the planning logic vs just clearly returning what version was used when creating the plan at the risk of more variables returned.